File: src/main/java/com/key/oa/entity/Permission.java
package com.key.oa.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.google.common.base.Objects;
import lombok.*;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

/**
 * @author 孙强
 * 资源的基类
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Permission {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * 资源的名称，便于用户理解
     */
    @Column(nullable = false)
    private String name;

    @Column(nullable = false, unique = true)
    private String value;

    @ManyToMany(mappedBy = "permissions")
    @ToString.Exclude
    @JsonIgnore
    private Set<Role> roles = new HashSet<>();

    /**
     * 这里的父子关系要求必须得删除掉所有的子角色后才能删除父角色（可以同时删除）
     */
    @ManyToOne(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private Permission parent;

    @ToString.Exclude
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "parent")
    private Set<Permission> children = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Permission that)) {
            return false;
        }
        return Objects.equal(id, that.id) && Objects.equal(value, that.value);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id, value);
    }
}

File: src/main/java/com/key/oa/entity/Permission.kt
package com.key.oa.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*

@Entity
class Permission(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY) var id: Long,

    @Column(nullable = false, unique = true) var value: String,

    @Column(nullable = false) var description: String,

    @JsonIgnore
    @ManyToMany(mappedBy = "permissions") var roles: MutableSet<Role> = HashSet(),

    @ManyToOne(cascade = [CascadeType.ALL], fetch = FetchType.EAGER)
    var parent: Permission?,

    @OneToMany(cascade = [CascadeType.ALL], mappedBy = "parent")
    var children: MutableSet<Permission> = HashSet(),
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Permission) return false

        if (id != other.id) return false
        if (value != other.value) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + value.hashCode()
        return result
    }
}